src/python/pants/engine/legacy/graph.py

```diff
@@ -360,24 +361,39 @@ def transitive_dependents_of_addresses(self, addresses):
         return result
 
 
-@dataclass(frozen=True)
+@frozen_after_init
+@dataclass
```

Comment:
Normally we use `unsafe_hash=True`, but here we define our own `__hash__` so that would be overridden.